www/js/lib/abstractFilesystemAccess.js

```diff
@@ -88,6 +89,216 @@ StorageFirefoxOS.prototype.enumerate = function (path) {
     return this._storage.enumerate();
 };
 
+// refer to this article for easy explanation of File System API https://developer.chrome.com/articles/file-system-access/
+
+/**
+ * @param {Array<string>} files All the File names to be shown in the dropdown
+ * @param {string} selectedFile The name of the file to be selected in the dropdown
+ * @returns {Promise<Array<string>>} Array of unique filenames (if a split zim is considered a single file)
+ */
+async function updateZimDropdownOptions (files, selectedFile) {
+    const select = document.getElementById('archiveList');
+    const options = [];
+    let count = 0;
+    if (files.length !== 0) options.push(new Option('Select an archive..', ''));
+
+    files.forEach((fileName) => {
+        if (fileName.endsWith('.zim') || fileName.endsWith('.zimaa')) {
+            options.push(new Option(`${fileName}`, fileName));
+            count++;
+        }
+    });
+    select.replaceChildren(...options);
+    document.getElementById('archiveList').value = selectedFile;
+    document.getElementById('numberOfFilesDisplay').innerText = count;
+    document.getElementById('fileCountDisplay').style.display = '';
+}
+
+/**
+ * Opens the File System API to select a directory
+ */
+async function selectDirectoryFromPickerViaFileSystemApi () {
+    const handle = await window.showDirectoryPicker();
+    const fileNames = [];
+    for await (const entry of handle.values()) {
+        fileNames.push(entry.name);
+    }
+
+    localStorage.setItem('zimFilenames', fileNames.join('|'));
+    updateZimDropdownOptions(fileNames, '');
+    cache.idxDB('zimFiles', handle, function () {
+        // save file in DB
+    });
+}
+
+/**
+ * Opens the File System API to select a file
+ * @returns {Promise<Array<File>>} The selected file from picker
+ */
+async function selectFileFromPickerViaFileSystemApi () {
+    const fileHandles = await window.showOpenFilePicker({ multiple: false });
+    const [selectedFile] = fileHandles;
+    const file = await selectedFile.getFile();
+
+    cache.idxDB('zimFiles', selectedFile, function () {
+        // file saved in DB
+    });
+    updateZimDropdownOptions([selectedFile.name], selectedFile.name);
+    return [file];
+}
+
+/**
+ * Gets the selected zim file from the IndexedDB
+ * @param {string} selectedFilename The name of the file to get back from DB
+ * @returns {Promise<Array<File>>} The selected File Object from cache
+ */
+function getSelectedZimFromCache (selectedFilename) {
+    return new Promise((resolve, _reject) => {
+        cache.idxDB('zimFiles', async function (fileOrDirHandle) {
+            console.log('[DEBUG] current file/dir permission status', await fileOrDirHandle.queryPermission(), fileOrDirHandle);
```

Comment:
I think this console.log should be removed before merge, as it adds some unnecessary async overhead (when compiled, it'll probably be a mass of code...).